File: Algorithms/135. Candy.rs
use std::cmp;

impl Solution {
    pub fn candy(ratings: Vec<i32>) -> i32 {
        let length = ratings.len();
        if length == 1 { return 1; }
        let mut res = 0;
        let mut left2right = vec![1; length];
        let mut right2left = vec![1; length];

        for i in 1..length {
            if ratings[i] > ratings[i - 1] {
                left2right[i] = left2right[i - 1] + 1;
            }
        }
        for i in (0..=(length - 2)).rev() {
            // println!("index: {}", i);
            if ratings[i] > ratings[i + 1] {
                right2left[i] = right2left[i + 1] + 1;
            }
        }
        for i in 0..length {
            res += cmp::max(left2right[i], right2left[i]);
        }

        res
    }
}

File: Algorithms/135. Candy.ts
function candy(ratings: number[]): number {
  const n = ratings.length;
  if (n === 1) {
    return 1;
  }
  let result = 0;
  const left2right = new Array(n).fill(1);
  const right2left = new Array(n).fill(1);

  for (let i = 1; i < n; i++) {
    if (ratings[i] > ratings[i - 1]) {
      left2right[i] = left2right[i - 1] + 1;
    }
  }

  for (let i = n - 2; i >= 0; i--) {
    if (ratings[i] > ratings[i + 1]) {
      right2left[i] = right2left[i + 1] + 1;
    }
  }
  for (let i = 0; i < n; i++) {
    result += Math.max(left2right[i], right2left[i]);
  }

  return result;
}
